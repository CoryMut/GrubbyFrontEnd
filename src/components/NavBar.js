import React, { useContext, useState } from "react";
import { UserContext } from "./UserContext";
import { logoutUser } from "../helpers/GrubbyAPI";
import { Link } from "react-router-dom";

import useMediaQuery from "@material-ui/core/useMediaQuery";

import { makeStyles } from "@material-ui/core/styles";
import AppBar from "@material-ui/core/AppBar";
import Toolbar from "@material-ui/core/Toolbar";
import Typography from "@material-ui/core/Typography";
import IconButton from "@material-ui/core/IconButton";
import MenuIcon from "@material-ui/icons/Menu";
import Drawer from "@material-ui/core/Drawer";
import Button from "@material-ui/core/Button";

import List from "@material-ui/core/List";
import ListItemIcon from "@material-ui/core/ListItemIcon";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import Logo from "../components/Logo";

import ArrowBackIcon from "@material-ui/icons/ArrowBack";
import HomeIcon from "@material-ui/icons/Home";
import CloudUploadIcon from "@material-ui/icons/CloudUpload";
import ExitToAppIcon from "@material-ui/icons/ExitToApp";
import CollectionsBookmarkIcon from "@material-ui/icons/CollectionsBookmark";
import PersonIcon from "@material-ui/icons/Person";
import FavoriteIcon from "@material-ui/icons/Favorite";
import SettingsIcon from "@material-ui/icons/Settings";

const drawerWidth = 240;

const useStyles = makeStyles((theme) => ({
    root: {
        flexGrow: 1,
    },
    menuButton: {
        marginRight: theme.spacing(2),
        color: "white",
    },
    title: {
        flexGrow: 1,
        textAlign: "center",
        [theme.breakpoints.down("xs")]: {
            marginRight: "20px",
        },
    },
    centerTitle: {
        textAlign: "none",
    },
    appBar: {
        backgroundColor: "#645579",
        zIndex: theme.zIndex.drawer + 1,
        justifyContent: "center",
        position: "sticky",
        top: "0px",
        left: "0px",
        height: "70px",
        [theme.breakpoints.down("md")]: {
            height: "64px",
        },
        borderBottom: "1px solid black",
    },
    drawer: {
        width: drawerWidth,
        flexShrink: 0,
    },
    drawerPaper: {
        width: drawerWidth,
    },
    drawerContainer: {
        overflow: "auto",
    },
    list: {
        width: "20vw",
        minWidth: 250,
    },
    listItem: {
        margin: "auto",
    },
    bottomItem: {
        position: "fixed",
        bottom: 0,
        width: drawerWidth,
    },
    button: {
        color: "white",
        textDecoration: "none",
        "&:hover": {
            color: "white",
        },
    },
}));

const NavBar = () => {
    const matches = useMediaQuery("(max-width:320px)");
    const hideButton = useMediaQuery("(max-width:700px)");

    const classes = useStyles();
    const { isLoggedIn, logout, isAdmin } = useContext(UserContext);
    const [menu, setMenu] = useState(false);

    const openMenu = () => {
        setMenu(true);
    };

    const closeMenu = () => {
        setMenu(false);
    };

    const handleLogout = async () => {
        await logoutUser();
        logout();
        closeMenu();
    };

    return (
        <div>
            <AppBar position="sticky" className={classes.appBar} color="inherit" elevation={1}>
                <Toolbar>
                    <IconButton edge="start" className={classes.menuButton} aria-label="menu" onClick={openMenu}>
                        <MenuIcon />
                    </IconButton>
                    <Typography variant="h6" className={matches ? classes.centerTitle : classes.title}>
                        <Logo></Logo>
                    </Typography>
                    {!isLoggedIn && !hideButton && (
                        <Button component={Link} to="/login" className={classes.button}>
                            Login
                        </Button>
                    )}
                    {isLoggedIn && !hideButton && (
                        <Button onClick={handleLogout} className={classes.button}>
                            Logout
                        </Button>
                    )}
                </Toolbar>
            </AppBar>
            <Drawer
                anchor="left"
                open={menu}
                onClose={closeMenu}
                className={classes.drawer}
                classes={{
                    paper: classes.drawerPaper,
                }}
            >
                <div className={classes.drawerContainer}>
                    <List>
                        <ListItem button onClick={closeMenu}>
                            <ListItemIcon>
                                <ArrowBackIcon></ArrowBackIcon>
                            </ListItemIcon>
                            <ListItemText></ListItemText>
                        </ListItem>

                        <ListItem button component={Link} to="/" onClick={closeMenu}>
                            <ListItemIcon>
                                <HomeIcon></HomeIcon>
                            </ListItemIcon>
                            <ListItemText primary="Home"></ListItemText>
                        </ListItem>

                        <ListItem button component={Link} to="/all?page=1" onClick={closeMenu}>
                            <ListItemIcon>
                                <CollectionsBookmarkIcon></CollectionsBookmarkIcon>
                            </ListItemIcon>
                            <ListItemText primary="All Comics"></ListItemText>
                        </ListItem>

                        {isLoggedIn && (
                            <ListItem button component={Link} onClick={closeMenu} to="/favorites">
                                <ListItemIcon>
                                    <FavoriteIcon></FavoriteIcon>
                                </ListItemIcon>
                                <ListItemText primary="Favorites"></ListItemText>
                            </ListItem>
                        )}

                        {!isLoggedIn && (
                            <ListItem
                                button
                                component={Link}
                                to="/login"
                                onClick={closeMenu}
                                className={classes.bottomItem}
                            >
                                <ListItemIcon>
                                    <PersonIcon></PersonIcon>
                                </ListItemIcon>
                                <ListItemText primary="Login"></ListItemText>
                            </ListItem>
                        )}

                        {isLoggedIn && (
                            <ListItem button onClick={handleLogout} className={classes.bottomItem}>
                                <ListItemIcon>
                                    <ExitToAppIcon></ExitToAppIcon>
                                </ListItemIcon>
                                <ListItemText primary="Logout"></ListItemText>
                            </ListItem>
                        )}

                        {isAdmin && (
                            <ListItem button component={Link} to="/upload" onClick={closeMenu}>
                                <ListItemIcon>
                                    <CloudUploadIcon></CloudUploadIcon>
                                </ListItemIcon>
                                <ListItemText primary="Upload"></ListItemText>
                            </ListItem>
                        )}
                        {isAdmin && (
                            <ListItem button component={Link} to="/admin" onClick={closeMenu}>
                                <ListItemIcon>
                                    <SettingsIcon></SettingsIcon>
                                </ListItemIcon>
                                <ListItemText primary="Admin Portal"></ListItemText>
                            </ListItem>
                        )}
                    </List>
                </div>
            </Drawer>
        </div>
    );
};

export default NavBar;
