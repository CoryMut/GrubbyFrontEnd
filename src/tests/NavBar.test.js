import React from "react";
import { shallow } from "enzyme";
import NavBar from "../components/NavBar";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router";

test("should test NavBar component", () => {
    const wrapper = shallow(<NavBar />);
    expect(wrapper).toMatchSnapshot();
});

test("it renders without crashing", () => {
    render(
        <MemoryRouter>
            <NavBar />
        </MemoryRouter>
    );
});
