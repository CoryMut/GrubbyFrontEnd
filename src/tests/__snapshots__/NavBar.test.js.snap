// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test NavBar component 1`] = `
<Fragment>
  <WithStyles(ForwardRef(AppBar))
    className="makeStyles-appBar-5"
    color="inherit"
    position="sticky"
  >
    <WithStyles(ForwardRef(Toolbar))>
      <WithStyles(ForwardRef(IconButton))
        aria-label="menu"
        className="makeStyles-menuButton-2"
        edge="start"
        onClick={[Function]}
      >
        <MenuIcon />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(Typography))
        className="makeStyles-title-3"
        variant="h6"
      >
        <Logo />
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Toolbar))>
  </WithStyles(ForwardRef(AppBar))>
  <WithStyles(ForwardRef(Drawer))
    anchor="left"
    className="makeStyles-drawer-6"
    classes={
      Object {
        "paper": "makeStyles-drawerPaper-7",
      }
    }
    onClose={[Function]}
    open={false}
  >
    <div
      className="makeStyles-drawerContainer-8"
    >
      <WithStyles(ForwardRef(List))>
        <WithStyles(ForwardRef(ListItem))
          button={true}
          onClick={[Function]}
        >
          <WithStyles(ForwardRef(ListItemIcon))>
            <ArrowBackIcon />
          </WithStyles(ForwardRef(ListItemIcon))>
          <WithStyles(ForwardRef(ListItemText)) />
        </WithStyles(ForwardRef(ListItem))>
        <WithStyles(ForwardRef(ListItem))
          button={true}
          component={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "Link",
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          onClick={[Function]}
          to="/"
        >
          <WithStyles(ForwardRef(ListItemIcon))>
            <HomeIcon />
          </WithStyles(ForwardRef(ListItemIcon))>
          <WithStyles(ForwardRef(ListItemText))
            primary="Home"
          />
        </WithStyles(ForwardRef(ListItem))>
        <WithStyles(ForwardRef(ListItem))
          button={true}
          component={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "Link",
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          onClick={[Function]}
          to="/all?page=1"
        >
          <WithStyles(ForwardRef(ListItemIcon))>
            <CollectionsBookmarkIcon />
          </WithStyles(ForwardRef(ListItemIcon))>
          <WithStyles(ForwardRef(ListItemText))
            primary="All Comics"
          />
        </WithStyles(ForwardRef(ListItem))>
        <WithStyles(ForwardRef(ListItem))
          button={true}
          className="makeStyles-bottomItem-11"
          component={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "Link",
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
          onClick={[Function]}
          to="/login"
        >
          <WithStyles(ForwardRef(ListItemIcon))>
            <PersonIcon />
          </WithStyles(ForwardRef(ListItemIcon))>
          <WithStyles(ForwardRef(ListItemText))
            primary="Login"
          />
        </WithStyles(ForwardRef(ListItem))>
      </WithStyles(ForwardRef(List))>
    </div>
  </WithStyles(ForwardRef(Drawer))>
</Fragment>
`;
